controller/controller_server/csi_controller_server.py

```diff
@@ -153,70 +156,69 @@ def CreateVolume(self, request, context):  # pylint: disable=too-many-branches
             context.set_details('an internal exception occurred : {}'.format(ex))
             return csi_pb2.CreateVolumeResponse()
 
-    def _copy_to_existing_volume_from_snapshot(self, vol, src_snapshot_id, min_vol_size, array_mediator, pool):
-        vol_name = vol.volume_name
+    def _copy_to_existing_volume_from_source_object(self, volume, source_object_id, source_object_type,
+                                                    minimum_volume_size, array_mediator, pool):
+        volume_name = volume.name
         try:
-            src_snapshot = array_mediator.get_snapshot_by_id(src_snapshot_id)
-            if not src_snapshot:
-                raise controller_errors.SnapshotNotFoundError(src_snapshot_id)
-            src_snapshot_name = src_snapshot.snapshot_name
-            src_snapshot_capacity = src_snapshot.capacity_bytes
-            logger.debug("Copy snapshot {0} data to volume {1}.".format(src_snapshot_id, vol_name))
-            array_mediator.copy_to_existing_volume_from_snapshot(vol_name, src_snapshot_name, src_snapshot_capacity,
-                                                                 min_vol_size, pool)
-            logger.debug("Copy volume from snapshot finished")
-        except controller_errors.VolumeNotFoundError as ex:
-            logger.error("Volume not found while copying snapshot data to volume")
+            source_object = array_mediator.get_object_by_id(source_object_id, source_object_type)
+            if not source_object:
+                self._rollback_create_volume_from_source(array_mediator, volume.id)
+                raise controller_errors.ObjectNotFoundError(source_object_id)
+            source_object_name = source_object.name
+            source_object_capacity = source_object.capacity_bytes
+            logger.debug("Copy {0} {1} data to volume {2}.".format(source_object_type, source_object_id, volume_name))
+            if source_object_type is config.OBJECT_TYPE_NAME_SNAPSHOT:
```

Comment:
shouldn't this be `VOLUME_SOURCE_SNAPSHOT`?

Reply:
not relevant

Reply:
doesn't have to be in this PR, but since you're already making changes here, just wondering if you agree that we should fix this in general